File: java-server/Java-Server/src/main/java/io/github/kawase/utility/EncryptionUtility.java
package io.github.kawase.utility;


import javax.crypto.Cipher;
import javax.crypto.SecretKey;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.SecureRandom;

public class EncryptionUtility {
    private static final String ALGORITHM = "AES/GCM/NoPadding";
    private static final int TAG_LENGTH_BIT = 128;
    private static final int IV_LENGTH_BYTE = 12;

    private static SecretKey deriveKey(final String password) throws Exception {
        final MessageDigest sha = MessageDigest.getInstance("SHA-256");
        byte[] key = sha.digest(password.getBytes(StandardCharsets.UTF_8));
        return new SecretKeySpec(key, "AES");
    }

    public static byte[] encryptLong(long value, final String key) throws Exception {
        byte[] bytes = new byte[Long.BYTES];
        for (int i = 7; i >= 0; i--) {
            bytes[i] = (byte) (value & 0xFF);
            value >>= 8;
        }

        return encryptBytes(bytes, key);
    }

    public static long decryptLong(final byte[] encryptedData, final String encryptionKey) throws Exception {
        byte[] decryptedBytes = decryptBytes(encryptedData, encryptionKey);

        if (decryptedBytes.length != Long.BYTES) {
            throw new IllegalArgumentException("Decrypted data length mismatch.");
        }

        long value = 0;
        for (int i = 0; i < Long.BYTES; i++) {
            value = (value << 8) | (decryptedBytes[i] & 0xFF);
        }

        return value;
    }

    public static byte[] encryptBytes(final byte[] data, final String encryptionKey) throws Exception {
        final byte[] iv = new byte[IV_LENGTH_BYTE];
        new SecureRandom().nextBytes(iv);

        final SecretKey key = deriveKey(encryptionKey);
        final Cipher cipher = Cipher.getInstance(ALGORITHM);
        cipher.init(Cipher.ENCRYPT_MODE, key, new GCMParameterSpec(TAG_LENGTH_BIT, iv));

        final byte[] cipherText = cipher.doFinal(data);

        return ByteBuffer.allocate(iv.length + cipherText.length)
                .put(iv)
                .put(cipherText)
                .array();
    }

    public static byte[] decryptBytes(final byte[] encryptedData, final String encryptionKey) throws Exception {
        final byte[] iv = new byte[IV_LENGTH_BYTE];
        System.arraycopy(encryptedData, 0, iv, 0, iv.length);

        final byte[] cipherText = new byte[encryptedData.length - IV_LENGTH_BYTE];
        System.arraycopy(encryptedData, IV_LENGTH_BYTE, cipherText, 0, cipherText.length);

        final SecretKey key = deriveKey(encryptionKey);
        final Cipher cipher = Cipher.getInstance(ALGORITHM);
        cipher.init(Cipher.DECRYPT_MODE, key, new GCMParameterSpec(TAG_LENGTH_BIT, iv));

        return cipher.doFinal(cipherText);
    }
}

File: unity/NeuroKey/Assets/Scripts/Runtime/Network/EncryptionUtility.cs
using System;
using System.IO;
using System.Security.Cryptography;
using System.Text;

namespace NeuroKey.Network
{
    public static class EncryptionUtility
    {
        private const int TagLengthBit = 128;
        private const int TagLengthByte = TagLengthBit / 8;
        private const int IvLengthByte = 12;

        private static byte[] DeriveKey(string password)
        {
            using (var sha = SHA256.Create())
            {
                return sha.ComputeHash(Encoding.UTF8.GetBytes(password));
            }
        }

        public static byte[] EncryptLong(long value, string key)
        {
            byte[] bytes = new byte[8];
            for (int i = 7; i >= 0; i--)
            {
                bytes[i] = (byte)(value & 0xFF);
                value >>= 8;
            }
            return EncryptBytes(bytes, key);
        }

        public static long DecryptLong(byte[] encryptedData, string encryptionKey)
        {
            byte[] decryptedBytes = DecryptBytes(encryptedData, encryptionKey);
            if (decryptedBytes.Length != 8)
                throw new ArgumentException("Decrypted data length mismatch.");

            long value = 0;
            for (int i = 0; i < 8; i++)
            {
                value = (value << 8) | (decryptedBytes[i] & 0xFF);
            }
            return value;
        }

        public static byte[] EncryptBytes(byte[] data, string encryptionKey)
        {
            byte[] key = DeriveKey(encryptionKey);
            byte[] iv = new byte[IvLengthByte];
            using (var rng = new RNGCryptoServiceProvider())
            {
                rng.GetBytes(iv);
            }

            byte[] tag = new byte[TagLengthByte];
            byte[] cipherText = new byte[data.Length];

            using (var aesGcm = new AesGcm(key))
            {
                aesGcm.Encrypt(iv, data, cipherText, tag);
            }

            byte[] finalBuffer = new byte[IvLengthByte + cipherText.Length + TagLengthByte];
            Buffer.BlockCopy(iv, 0, finalBuffer, 0, IvLengthByte);
            Buffer.BlockCopy(cipherText, 0, finalBuffer, IvLengthByte, cipherText.Length);
            Buffer.BlockCopy(tag, 0, finalBuffer, IvLengthByte + cipherText.Length, TagLengthByte);

            return finalBuffer;
        }

        public static byte[] DecryptBytes(byte[] encryptedData, string encryptionKey)
        {
            byte[] key = DeriveKey(encryptionKey);
            byte[] iv = new byte[IvLengthByte];
            Buffer.BlockCopy(encryptedData, 0, iv, 0, IvLengthByte);

            int cipherTextLength = encryptedData.Length - IvLengthByte - TagLengthByte;
            byte[] cipherText = new byte[cipherTextLength];
            byte[] tag = new byte[TagLengthByte];

            Buffer.BlockCopy(encryptedData, IvLengthByte, cipherText, 0, cipherTextLength);
            Buffer.BlockCopy(encryptedData, IvLengthByte + cipherTextLength, tag, 0, TagLengthByte);

            byte[] decryptedData = new byte[cipherTextLength];

            using (var aesGcm = new AesGcm(key))
            {
                aesGcm.Decrypt(iv, cipherText, tag, decryptedData);
            }

            return decryptedData;
        }
    }
}
